kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.provider.hashicorp.vault;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.kroxylicious.kms.provider.hashicorp.vault.VaultResponse.DataKeyData;
+import io.kroxylicious.kms.provider.hashicorp.vault.VaultResponse.DecryptData;
+import io.kroxylicious.kms.provider.hashicorp.vault.VaultResponse.ReadKeyData;
+import io.kroxylicious.kms.service.DekPair;
+import io.kroxylicious.kms.service.Kms;
+import io.kroxylicious.kms.service.KmsException;
+import io.kroxylicious.kms.service.Serde;
+import io.kroxylicious.kms.service.UnknownKeyException;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * {@inheritDoc}
+ * <br/>
+ * An implementation of the KMS interface backed by a remote instance of HashiCorp Vault (v1).
+ * <br/>
+ * <h2>TODO</h2>
+ * <ul>
+ *    <li>don't assume /transit endpoint</li>
+ *    <li>HTTPs</li>
+ *    <li>Client trust store</li>
+ *    <li>Securely pass vault token</li>
+ * </ul>
+ */
+public class VaultKms implements Kms<String, VaultEdek> {
+
+    private static final String AES_KEY_ALGO = "AES";
+
+    private final HttpClient vaultClient = HttpClient.newBuilder()
+            .followRedirects(HttpClient.Redirect.NORMAL)
+            .connectTimeout(Duration.ofSeconds(20))
+            .build();
+    private final URI vaultUrl;
+    private final String vaultToken;
+
+    VaultKms(URI vaultUrl, String vaultToken) {
+        this.vaultUrl = vaultUrl;
+        this.vaultToken = vaultToken;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br/>
+     * @see <a href="https://developer.hashicorp.com/vault/api-docs/secret/transit#generate-data-key">https://developer.hashicorp.com/vault/api-docs/secret/transit#generate-data-key</a>
+     */
+    @NonNull
+    @Override
+    public CompletionStage<DekPair<VaultEdek>> generateDekPair(@NonNull String kekRef) {
+
+        var request = createVaultRequest()
+                .uri(vaultUrl.resolve("v1/transit/datakey/plaintext/%s".formatted(kekRef)))
+                .POST(HttpRequest.BodyPublishers.noBody())
+                .build();
+
+        return vaultClient.sendAsync(request, statusHandler(kekRef, new JsonBodyHandler<VaultResponse<DataKeyData>>(new TypeReference<>() {
```

Comment:
What thread does this sendAsync complete on? Because IIUC that details leaks to the caller of `generateDekPair`, who might need to volatile/synchronized etc to make their code threadsafe.

Reply:
Let's add a sentence to the Javadoc of the `Kms` to say that the thread on which stages complete is undefined. That way at least the contract is explicit.